cpp/src/arrow/filesystem/path_util.h

```diff
@@ -113,6 +123,27 @@ std::string JoinAbstractPath(const StringRange& range) {
   return JoinAbstractPath(range.begin(), range.end());
 }
 
+// Join the components of filename partitions
+template <class StringIt>
+std::string JoinFilenamePartitions(StringIt it, StringIt end) {
```

Comment:
Same concern here.  This class is meant to be more generic so we can't refer to things like "partitions" and `_` which are concepts that only make sense within the domain of `partitioning.h`.

Perhaps move this method inside of `partitioning.cc`?  Or you could make a generic `JoinPaths(const StringRange& range, const std::string& sep)` but I wouldn't expect it to put `sep` at the end of the path.

Reply:
Used a generic Join function and concatenated the separator in the end of the path for FilenamePartitioning in partition.cc